tensorlayer/optimizers/lazyAdam.py

```diff
@@ -0,0 +1,147 @@
+        """Applies grad for one step."""
+        var_dtype = var.dtype.base_dtype
+        lr_t = self._decayed_lr(var_dtype)
+        beta_1_t = self._get_hyper('beta_1', var_dtype)
+        beta_2_t = self._get_hyper('beta_2', var_dtype)
+        local_step = tf.cast(self.iterations + 1, var_dtype)
+        beta_1_power = tf.math.pow(beta_1_t, local_step)
+        beta_2_power = tf.math.pow(beta_2_t, local_step)
+        epsilon_t = tf.convert_to_tensor(self.epsilon, var_dtype)
+        lr = (lr_t * tf.math.sqrt(1 - beta_2_power) / (1 - beta_1_power))
+
+        # \\(m := beta1 * m + (1 - beta1) * g_t\\)
+        m = self.get_slot(var, 'm')
+        m_t_slice = beta_1_t * tf.gather(m, indices) + (1 - beta_1_t) * grad
+
+        m_update_kwargs = {'resource': m.handle, 'indices': indices, 'updates': m_t_slice}
+        m_update_op = tf.raw_ops.ResourceScatterUpdate(**m_update_kwargs)
+
+        # \\(v := beta2 * v + (1 - beta2) * (g_t * g_t)\\)
+        v = self.get_slot(var, 'v')
+        v_t_slice = (beta_2_t * tf.gather(v, indices) + (1 - beta_2_t) * tf.math.square(grad))
+
+        v_update_kwargs = {'resource': v.handle, 'indices': indices, 'updates': v_t_slice}
+        v_update_op = tf.raw_ops.ResourceScatterUpdate(**v_update_kwargs)
+
+        # \\(variable -= learning_rate * m_t / (epsilon_t + sqrt(v_t))\\)
+        var_slice = lr * m_t_slice / (tf.math.sqrt(v_t_slice) + epsilon_t)
+
+        var_update_kwargs = {'resource': var.handle, 'indices': indices, 'updates': var_slice}
+        var_update_op = tf.raw_ops.ResourceScatterSub(**var_update_kwargs)
+
+        return tf.group(*[var_update_op, m_update_op, v_update_op])
+
+
+class LearningRateFn(object):
+    """Creates learning rate function."""
+
+    def __init__(self, learning_rate, hidden_size, warmup_steps):
+        self.learning_rate = learning_rate
+        self.hidden_size = hidden_size
+        self.warmup_steps = float(warmup_steps)
+
+    def __call__(self, global_step):
+        """Calculate learning rate with linear warmup and rsqrt decay."""
+        step = float(global_step)
+        learning_rate = self.learning_rate
+        learning_rate *= (self.hidden_size**-0.5)
+        # Apply linear warmup
+        learning_rate *= np.minimum(1.0, step / self.warmup_steps)
+        # Apply rsqrt decay
+        learning_rate /= np.sqrt(np.maximum(step, self.warmup_steps))
+        return learning_rate
+
+
+class LearningRateScheduler(tf.keras.callbacks.Callback):
+    """Keras callback to schedule learning rate.
+
+  TODO(tianlin): Refactor this scheduler and LearningRateBatchScheduler in
+  official/resnet/keras/keras_common.py.
+  """
+
+    def __init__(self, schedule, init_steps=None, verbose=False):
+        super(LearningRateScheduler, self).__init__()
+        self.schedule = schedule
+        self.verbose = verbose
+        if init_steps is None:
+            init_steps = 0.0
+        self.steps = float(init_steps)  # Total steps during training.
```

Comment:
Why steps is a float not an int?

Reply:
I will get rid of this part. It has been updated in tf.keras library. thanks